### PR TITLE
Add comparison maps with WOA18 climatology

### PR DESCRIPTION
**Pull request by milenaveneziani**
This is a first attempt at adding comparison maps with WOA18 climatological data.

WOA18 is an updated version with respect to WOA13 in that it includes polar observations, making it more suitable for Arctic and Southern Ocean comparisons.
The data I downloaded has 0.25deg resolution and is available as annual climatologies over depths 0-5500 m, and as monthly climatologies over depths 0-1500 m.

### CODE REVIEW

**Comment by milenaveneziani** [comment]
@xylar, I am getting a task assertion error when I ask for more than 1 season (say `seasons=['ANN','JFM']`). Either `ANN` or `JFM` works by itself, but not together. Do you understand where the problem might be?
Here is the full error:
```
Traceback (most recent call last):
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.6/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/global/project/projectdirs/acme/software/anaconda_envs/cori/base/envs/e3sm_unified_1.2.6/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/__main__.py", line 862, in <module>
    main()
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/__main__.py", line 843, in main
    analyses = determine_analyses_to_generate(analyses, args.verbose)
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/__main__.py", line 267, in determine_analyses_to_generate
    verbose)
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/__main__.py", line 376, in add_task_and_subtasks
    assert(failureCount > 0)
AssertionError
```

**Comment by milenaveneziani** [comment]
Also, I think I added or updated all the necessary files related to new obs. But please let me know if there is anything I missed (I still need to add a png example, but it's on my list).
Could you remind me how I can generate an updated documentation?

**Comment by xylar** [comment]
> I am getting a task assertion error when I ask for more than 1 season (say seasons=['ANN','JFM']). Either ANN or JFM works by itself, but not together. Do you understand where the problem might be?

This is just a hunch, but maybe you have 2 subtasks with the same name.  That will really mess things up.  It might be that your subtasks need to include the name of the season. If you don't see anything obvious, I'll try to run myself as soon as I can, maybe sometime over the weekend.

**Comment by xylar** [comment]
> Also, I think I added or updated all the necessary files related to new obs. But please let me know if there is anything I missed (I still need to add a png example, but it's on my list).
Could you remind me how I can generate an updated documentation?

You should be able to do:
```
cd docs
conda env create -f environment.yml
conda activate mpas-analysis-docs
make html
```
After that, look in `_build/html`

**Comment by milenaveneziani** [comment]
> This is just a hunch, but maybe you have 2 subtasks with the same name.

Indeed! Fixed it and it worked, thanks.
I'll be pushing a couple more times and then this is ready to be tested.

**Comment by milenaveneziani** [comment]
OK, I am done for now. This is ready to be tested.

Please note that, for some reason, I could not run the `make html` command. I get a `ModuleNotFoundError: No module named 'mpas_analysis'` error. So, if you could please check how the documentation looks like, I'd appreciate it.
(Not sure why it can't find the module: I did deactivate my old conda env and followed your directions)

**Comment by milenaveneziani** [comment]
Also, I have updated the public repo on anvil (files are also available on the diagnostics folder on NERSC and LANL).

**Comment by xylar** [comment]
Sorry @milenaveneziani.  Here's what I actually run to test the docs:
```bash
# test building the docs
conda remove -y --all -n mpas-analysis-docs
conda env create -f docs/environment.yml
conda install -y -n mpas-analysis-docs -c defaults mock pillow sphinx sphinx_rtd_theme
conda activate mpas-analysis-docs
pip install .
rm -rf build dist mpas_analysis.egg-info
cd docs
make clean
make html
```
Clearly, it's more complicated than I remembered.  Maybe easier to leave it to me.

**Comment by milenaveneziani** [comment]
OK, I did all those steps and all went well. The only thing left is to visualize the html page :)

**Comment by milenaveneziani** [comment]
Updated documentation can be seen here:
https://portal.nersc.gov/project/acme/milena/mpas-analysis-doc/

**Comment by xylar** [comment]
I'm reviewing this PR now.  Could you add the new data sets, readme and bib to this file?
https://github.com/milenaveneziani/MPAS-Analysis/blob/add_WOA18comparison/mpas_analysis/obs/analysis_input_files#L100
That way I can use the download script to get the data from Anvil.

**Comment by xylar** [comment]
Also, permissions need to chance on Anvil so the files are available for download:
https://web.lcrc.anl.gov/public/e3sm/diagnostics/observations/Ocean/WOA18/

**Comment by xylar** [comment]
One more thing about that directory.  Could you just rename the files to have the date at the end, rather than having symlinks?  The only reason we have symlinks in other directories is that older versions of MPAS-Analysis don't know to look for the files with the dates on the end, but going forward we want all NetCDF files in observations to have a date.  (It would be good if mapping files and masks did, too, in case we need to update them, but we're not there yet.)

File: mpas_analysis/ocean/plot_climatology_map_subtask.py [hunk 1]
```diff
@@ -550,6 +553,77 @@ def _plot_antarctic(self, remappedModelClimatology,
...
+            groupSubtitle=self.groupSubtitle,
+            groupLink=self.groupLink,
+            gallery=self.galleryName,
+            thumbnailDescription=self.thumbnailDescription,
+            imageDescription=caption,
+            imageCaption=caption)
+
+        # }}}
+
+    def _plot_arctic(self, remappedModelClimatology,
```
Comment (xylar) [inline mpas_analysis/ocean/plot_climatology_map_subtask.py]: @milenaveneziani, would it make sense to combine `_plot_antarctic()` and `_plot_arctic()` into a single function?  They seem nearly identical.

File: mpas_analysis/shared/plot/climatology_map.py [hunk 2]
```diff
@@ -403,9 +403,6 @@ def plot_polar_projection_comparison(
     title : str, optional
         the subtitle of the plot
 
-    plotProjection : str, optional
-        Basemap projection for the plot
-
```
Comment (xylar) [inline mpas_analysis/shared/plot/climatology_map.py]: Oh, nice catch!

**Comment by milenaveneziani** [comment]
> Could you just rename the files to have the date at the end, rather than having symlinks?

oh so the symlink should go the other way, right? I remember this from past discussions, but it slipped my mind.

File: mpas_analysis/ocean/plot_climatology_map_subtask.py (hunk 1 above)
Comment (milenaveneziani) [reply mpas_analysis/ocean/plot_climatology_map_subtask.py]: yes, I noticed that too. How about `_plot_polar` or maybe `_plot_polar_stereographic`?

**Comment by xylar** [comment]
> oh so the symlink should go the other way, right? I remember this from past discussions, but it slipped my mind.

We don't want the symlinks at all.  We just want the dates to be there in all files.  Again, the versions without dates are only there for existing observations for backwards compatibility.  From here on, we only want files with dates, none without.

**Comment by xylar** [comment]
The code is currently generating two mapping files, one for T and one for S even though these come from the same file and the same grid.  I'm going to try to figure out if there's a way to avoid this, because we only need one mapping file.

**Comment by xylar** [comment]
Hmm, it seems like we have that redundancy for the Argo data, too.  So maybe easiest to leave it as it is, especially since the mapping files are already in place on Anvil.

Speaking of which, they also need to be added to 
https://github.com/milenaveneziani/MPAS-Analysis/blob/add_WOA18comparison/mpas_analysis/obs/analysis_input_files

File: mpas_analysis/ocean/plot_climatology_map_subtask.py (hunk 1 above)
Comment (xylar) [reply mpas_analysis/ocean/plot_climatology_map_subtask.py]: Yep, either of those if perfect.

**Comment by xylar** [comment]
My test of just the new `climatologyMapWoa` task worked great on a QU240 simulation.  I'm testing now with all 3 comparison grids ('latlon', 'antarctic' and 'arctic') and also with the Argo task using all 3 grids.

Then, I'll run a full stress test: py2, py3 and main vs. control runs.  I'll let you know how that goes.

**Comment by milenaveneziani** [comment]
great, thanks.

I think I have done everything you suggested, except for making that one _plot_polar function. Any preference on the name I could use?

**Comment by xylar** [comment]
Let's go with just `_plot_polar`.  I don't think there's any room for confusion about what that means.

**Comment by xylar** [comment]
Yep, Argo was great!  Even the gallery names are right!  Looks terrible in polar regions, unfortunately so we'd never actually want to plot those.  But it proves that the new Arctic grid could potentially be used for any task that supports multiple comparison grids.

On to the stress tests...

**Comment by xylar** [comment]
One more small thing, could you remove the whitespace that PEP8 speaks is complaining about?
https://github.com/MPAS-Dev/MPAS-Analysis/pull/622#issuecomment-525895836

**Comment by milenaveneziani** [comment]
so, I can't actually see those white spaces.. I'm not sure why it is complaining.

**Comment by milenaveneziani** [comment]
Could you also use the download script to make sure that permissions are OK?

I will also do a giant squash before merging this.

**Comment by milenaveneziani** [comment]
About this:
> Line 456:14: E127 continuation line over-indented for visual indent

I actually like it as is, if you are OK with it.

**Comment by xylar** [comment]
> Line 456:14: E127 continuation line over-indented for visual indent

I think that one was already fixed before.  It's only complaining about spaces on blank lines at this point.  If you can't seen them, I'll fix them but they're definitely there.

**Comment by milenaveneziani** [comment]
no, line 456 has the updated elif statement, and it's got an extra space in it.

Curious about how you removed the invisible extra spaces?

**Comment by xylar** [comment]
> no, line 456 has the updated elif statement, and it's got an extra space in it.

I see, I agree, that's fine as it is.

> Curious about how you removed the invisible extra spaces?

I have spyder configured to remove white space at the end of lines every time I save.  But you can see that the cursor can move an extra space in vim or whatever editor you use.  You can use this trick in vim: https://vi.stackexchange.com/a/456

**Comment by milenaveneziani** [comment]
> But you can see that the cursor can move an extra space in vim or whatever editor you use. You can use this trick in vim: https://vi.stackexchange.com/a/456

yeah, without the trick I couldn't see the cursor moving anywhere.
Anyway, great. Let me know if the download script works fine and how the stress tests go. thanks.

**Comment by xylar** [comment]
The download script worked great, so permissions seem fine.  I'm having trouble building the docs -- conda is complaining about conflicts.  Odd because this worked previously.  I'll keep you posted.

**Comment by milenaveneziani** [comment]
> I'm having trouble building the docs

it worked for me, starting from the latest e3sm-unified and followed all your steps above.
BTW: do you want to update these steps in the quick start guide?

**Comment by xylar** [comment]
> do you want to update these steps in the quick start guide?

That's a good idea.  I'll push an updated set of instructions as soon as I get things working for me.

**Comment by xylar** [comment]
Okay, I fixed a few small things related to the docs and updated the `README.md`.  Trying the major stress tests again...

**Comment by xylar** [comment]
@milenaveneziani, all my tests passed (QU240wLI simulation with py2.7 and py3.7, as well as a main vs. control).  If you want to squash commits, go for it.  Then, you can feel free to hit "Merge pull request" and paste in the message from the top of the PR.  Please leave the PR number in the title of the commit message -- this helps me find which PRs are part of each release.

**Comment by milenaveneziani** [comment]
Thanks @xylar.

>  Please leave the PR number in the title of the commit message

yes,  I got used to doing this for any git merges I do :)

**Comment by milenaveneziani** [comment]
Thanks @xylar. I appreciate the help on this one. I hope we can use this data set effectively for other tasks as well. One logical example is to do transects, for instance.

**Comment by xylar** [comment]
>  I hope we can use this data set effectively for other tasks as well. 

For sure!  I plan to use it in #616.

> One logical example is to do transects, for instance.

Yeah, definitely.  The main thing would be to define the transects (similar to SOSE transects we currently have).

**Comment by xylar** [comment]
We still need to propagate the observations and mapping files to the supported machines.  Let me take care of that.  I have a good set of scripts for taking care of it.

**Comment by milenaveneziani** [comment]
I have it on cori and IC (just need to change the symlink thing: let me take care of it now).

**Comment by xylar** [comment]
@milenaveneziani, one thing that occurs to me.  WOA18 is in situ temperature, right?  Don't we need to use `gsw` to convert to potential temperature before plotting?  Or has this been done elsewhere?  If so, do we need to change the metadata in the NetCDF files to indicate this?

**Comment by milenaveneziani** [comment]
@xylar: I will need to check on this. A first quick read of their WOA18 documentation for Temperature did not reveal anything, so I may have to dig deeper.

**Pull request merged**